presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -0,0 +1,130 @@
+    private boolean isDeleteQuery;
+
+    public DetermineJoinDistributionType(CostComparator costComparator)
+    {
+        this.costComparator = costComparator;
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        if (node instanceof DeleteNode) {
+            isDeleteQuery = true;
+            return Optional.empty();
+        }
+        if (!(node instanceof JoinNode || node instanceof SemiJoinNode)) {
+            return Optional.empty();
+        }
+
+        return Optional.of(chooseJoinTypeFor(node, lookup, session, symbolAllocator));
+    }
+
+    private PlanNode chooseJoinTypeFor(PlanNode node, Lookup lookup, Session session, SymbolAllocator symbolAllocator)
+    {
+        Ordering<PlanNode> planNodeOrdering = new Ordering<PlanNode>()
+        {
+            @Override
+            public int compare(PlanNode node1, PlanNode node2)
+            {
+                PlanNodeCostEstimate node1Cost = lookup.getCumulativeCost(node1, session, symbolAllocator.getTypes());
+                PlanNodeCostEstimate node2Cost = lookup.getCumulativeCost(node2, session, symbolAllocator.getTypes());
+                return costComparator.compare(session, node1Cost, node2Cost);
+            }
+        };
+
+        FeaturesConfig.JoinDistributionType joinDistributionType = getJoinDistributionType(session);
+
+        if (node instanceof  JoinNode) {
+            return chooseBestTypeForJoin((JoinNode) node, planNodeOrdering, joinDistributionType, lookup);
+        }
+        else {
+            return chooseBestTypeForSemiJoin((SemiJoinNode) node, planNodeOrdering, joinDistributionType);
+        }
+    }
+
+    private JoinNode chooseBestTypeForJoin(JoinNode node, Ordering<PlanNode> planNodeOrdering, FeaturesConfig.JoinDistributionType joinDistributionType, Lookup lookup)
+    {
+        List<JoinNode> possibleJoinNodes = new ArrayList<>();
+
+        if (joinDistributionType.canRepartition() && !mustBroadcastJoin(node, lookup)) {
+            possibleJoinNodes.add(node.withDistributionType(JoinNode.DistributionType.PARTITIONED));
+            possibleJoinNodes.add(node.flipChildren().withDistributionType(JoinNode.DistributionType.PARTITIONED));
+        }
+
+        // The implementation of full outer join only works if the data is hash partitioned. See LookupJoinOperators#buildSideOuterJoinUnvisitedPositions
+        if (joinDistributionType.canReplicate() && node.getType() != FULL && node.getType() != RIGHT) {
+            possibleJoinNodes.add(node.withDistributionType(JoinNode.DistributionType.REPLICATED));
+            possibleJoinNodes.add(node.flipChildren().withDistributionType(JoinNode.DistributionType.REPLICATED));
+        }
+
+        checkState(!possibleJoinNodes.isEmpty(), "No valid join type found");
+        return planNodeOrdering.min(possibleJoinNodes);
```

Comment:
if cost are unknown the cost comparator will throw an error.  
Before comparing you should check for each !cost.isUnknown() and if any of them has an unknown cost return the required distribution type without flipping children (or if the type is automatic and there are no other constraints, return repartitioned).  

For ReorderJoins I ended up creating a JoinEnumerationResult to keep the cost and plans as a single object  so we would only compute the cost once.  And then instead of a planNodeOrdering, I had a JoinEnumerationResultOrdering. Something similar might be useful here.